Biomedical text natural language processing (BioNLP) using scispaCy


How to identify diseases, drugs, and dosages from medical record transcriptions (https://gbnegrini.com/post/biomedical-text-nlp-scispacy-named-entity-recognition-medical-records/)

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
!conda install -c conda-forge spacy
!conda install -c conda-forge scispacy
!-m spacy download en_core_sci_md (SciSpacy)

/bin/bash: conda: command not found
/bin/bash: conda: command not found
/bin/bash: - : invalid option
Usage:	/bin/bash [GNU long option] [option] ...
	/bin/bash [GNU long option] [option] script-file ...
GNU long options:
	--debug
	--debugger
	--dump-po-strings
	--dump-strings
	--help
	--init-file
	--login
	--noediting
	--noprofile
	--norc
	--posix
	--pretty-print
	--rcfile
	--restricted
	--verbose
	--version
Shell options:
	-ilrsD or -c command or -O shopt_option		(invocation only)
	-abefhkmnptuvxBCHP or -o option


In [ ]:
!pip install -U spacy 
!pip install scispacy 
!pip install https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.2.4/en_core_sci_sm-0.2.4.tar.gz 
!pip install https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.2.4/en_ner_craft_md-0.2.4.tar.gz 
!pip install https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.2.4/en_ner_jnlpba_md-0.2.4.tar.gz
!pip install https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.2.4/en_ner_bc5cdr_md-0.2.4.tar.gz
!pip install https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.2.4/en_ner_bionlp13cg_md-0.2.4.tar.gz 
!pip install https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.2.4/en_core_sci_lg-0.2.4.tar.gz

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Using cached spacy-3.5.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (6.6 MB)
  Attempting uninstall: spacy
    Found existing installation: spacy 3.4.4
    Uninstalling spacy-3.4.4:
      Successfully uninstalled spacy-3.4.4
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
scispacy 0.5.2 requires spacy<3.5.0,>=3.4.0, but you have spacy 3.5.2 which is incompatible.
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Using cached spacy-3.4.4-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (6.5 MB)
  Attempting uninstall: spacy
    Found existing installation: spacy 3.5.2
    Uninstalling spacy-3.5.2:
      Successfully uninstalled spacy-3.5.2
ERROR: pip's dependency resolver does not curren

In [ ]:
#Source Code
# Load Pkgs
import numpy as np # linear algebra
import pandas as pd # dataframe processing
import spacy
#scispaCy is a Python package containing spaCy models for processing biomedical, scientific or clinical text.
#https://allenai.github.io/scispacy/
import scispacy

# Text Preprocessing Pkg
from spacy.lang.en.stop_words import STOP_WORDS # stop words list
from string import punctuation
from collections import Counter
from string import punctuation
from spacy.tokenizer import Tokenizer
#calling a pretrained biomedical model 
#Assigns context-specific token vectors, POS tags, dependency parse and named entities.
#A full spaCy pipeline for biomedical data with a larger vocabulary and 50k word vectors.

#nlp = spacy.load("en_core_sci_md") #the built-in load function.

#NER specific models
#https://towardsdatascience.com/using-scispacy-for-named-entity-recognition-785389e7918d
import en_ner_craft_md
import en_ner_bc5cdr_md
import en_ner_jnlpba_md
import en_ner_bionlp13cg_md

#Tools for extracting & displaying data

#text post processing
from spacy import displacy   #for ner   

In [ ]:
!pip install https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.4.0/en_ner_bc5cdr_md-0.4.0.tar.gz

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Using cached https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.4.0/en_ner_bc5cdr_md-0.4.0.tar.gz (125.1 MB)
  Preparing metadata (setup.py) ... done
  Using cached spacy-3.0.9-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (5.9 MB)
  Using cached thinc-8.0.17-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (659 kB)
  Created wheel for en-ner-bc5cdr-md: filename=en_ner_bc5cdr_md-0.4.0-py3-none-any.whl size=125666879 sha256=a676a460daa4b83de66407505ce0b4ac5d14aaf0e8315e16ce720e8496f20520
  Stored in directory: /root/.cache/pip/wheels/c3/f5/32/313d08b812c91abeb6fb1d3b0f8fd69687c30c3a9d38288e4c
Successfully built en-ner-bc5cdr-md
  Attempting uninstall: thinc
    Found existing installation: thinc 8.1.10
    Uninstalling thinc-8.1.10:
      Successfully uninstalled thinc-8.1.10
  Attempting uninstall: spacy
    Found existing installation: spacy 3.4.4
   

In [ ]:
import pandas as pd

med_transcript = pd.read_csv("/content/drive/MyDrive/mtsamples.csv", index_col=0)
med_transcript.info()
med_transcript.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4999 entries, 0 to 4998
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   description        4999 non-null   object
 1   medical_specialty  4999 non-null   object
 2   sample_name        4999 non-null   object
 3   transcription      4966 non-null   object
 4   keywords           3931 non-null   object
dtypes: object(5)
memory usage: 234.3+ KB


description  \
0   A 23-year-old white female presents with comp...   
1           Consult for laparoscopic gastric bypass.   
2           Consult for laparoscopic gastric bypass.   
3                             2-D M-Mode. Doppler.     
4                                 2-D Echocardiogram   

             medical_specialty                                sample_name  \
0         Allergy / Immunology                         Allergic Rhinitis    
1                   Bariatrics   Laparoscopic Gastric Bypass Consult - 2    
2                   Bariatrics   Laparoscopic Gastric Bypass Consult - 1    
3   Cardiovascular / Pulmonary                    2-D Echocardiogram - 1    
4   Cardiovascular / Pulmonary                    2-D Echocardiogram - 2    

                                       transcription  \
0  SUBJECTIVE:,  This 23-year-old white female pr...   
1  PAST MEDICAL HISTORY:, He has difficulty climb...   
2  HISTORY OF PRESENT ILLNESS: , I have seen ABC ...   
3  2-D M-MODE: , ,1.  Left atrial enlargement wit...   
4  1.  The left ventricular cavity size and wall ...   

                                            keywords  
0  allergy / immunology, allergic rhinitis, aller...  
1  bariatrics, laparoscopic gastric bypass, weigh...  
2  bariatrics, laparoscopic gastric bypass, heart...  
3  cardiovascular / pulmonary, 2-d m-mode, dopple...  
4  cardiovascular / pulmonary, 2-d, doppler, echo...

The dataset has almost 5000 records, but let’s work with a small random subsample so it doesn’t take too long to process. We also have to drop any rows whose transcriptions are missing.

In [ ]:
med_transcript.dropna(subset=['transcription'], inplace=True)
med_transcript_small = med_transcript.sample(n=100, replace=False, random_state=42)
med_transcript_small.info()
med_transcript_small.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 100 entries, 3162 to 3581
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   description        100 non-null    object
 1   medical_specialty  100 non-null    object
 2   sample_name        100 non-null    object
 3   transcription      100 non-null    object
 4   keywords           78 non-null     object
dtypes: object(5)
memory usage: 4.7+ KB


description  \
3162   Markedly elevated PT INR despite stopping Cou...   
1981   Intercostal block from fourth to tenth interc...   
1361   The patient is a 65-year-old female who under...   
3008   Construction of right upper arm hemodialysis ...   
4943   Bronchoscopy with brush biopsies.  Persistent...   

                   medical_specialty                          sample_name  \
3162           Hematology - Oncology              Hematology Consult - 1    
1981                 Pain Management               Intercostal block - 1    
1361   SOAP / Chart / Progress Notes                Lobectomy - Followup    
3008                      Nephrology   Hemodialysis Fistula Construction    
4943      Cardiovascular / Pulmonary                    Bronchoscopy - 8    

                                          transcription  \
3162  HISTORY OF PRESENT ILLNESS:,  The patient is w...   
1981  PREPROCEDURE DIAGNOSIS:,  Chest pain secondary...   
1361  HISTORY OF PRESENT ILLNESS: , The patient is a...   
3008  PREOPERATIVE DIAGNOSIS: , End-stage renal dise...   
4943  PREOPERATIVE DIAGNOSIS: , Persistent pneumonia...   

                                               keywords  
3162                                                NaN  
1981  pain management, xylocaine, marcaine, intercos...  
1361  soap / chart / progress notes, non-small cell ...  
3008  nephrology, end-stage renal disease, av dialys...  
4943  cardiovascular / pulmonary, persistent pneumon...

In [ ]:
sample_transcription = med_transcript_small['transcription'].iloc[0]
print(sample_transcription[:1000]) # prints just the first 1000 characters

HISTORY OF PRESENT ILLNESS:,  The patient is well known to me for a history of iron-deficiency anemia due to chronic blood loss from colitis.  We corrected her hematocrit last year with intravenous (IV) iron.  Ultimately, she had a total proctocolectomy done on 03/14/2007 to treat her colitis.  Her course has been very complicated since then with needing multiple surgeries for removal of hematoma.  This is partly because she was on anticoagulation for a right arm deep venous thrombosis (DVT) she had early this year, complicated by septic phlebitis.,Chart was reviewed, and I will not reiterate her complex history.,I am asked to see the patient again because of concerns for coagulopathy.,She had surgery again last month to evacuate a pelvic hematoma, and was found to have vancomycin resistant enterococcus, for which she is on multiple antibiotics and followed by infectious disease now.,She is on total parenteral nutrition (TPN) as well.,LABORATORY DATA:,  Labs today showed a white blood 

Named entity recognition
Named entity recognition (NER) is a subtask of natural language processing used to identify and classify named entities mentioned in unstructured text into pre-defined categories. scispaCy has different models to identify different entity types and you can check them here.

We are going to use the NER model trained on the BC5CDR corpus (en_ner_bc5cdr_md). This corpus consists of 1500 PubMed articles with 4409 annotated chemicals, 5818 diseases, and 3116 chemical-disease interactions. Don’t forget to download and install the model.

In [ ]:
import scispacy
import spacy
nlp = spacy.load("en_ner_bc5cdr_md")

/usr/local/lib/python3.10/dist-packages/spacy/util.py:877: UserWarning: [W095] Model 'en_ner_bc5cdr_md' (0.4.0) was trained with spaCy v3.0 and may not be 100% compatible with the current version (3.4.4). If you see errors or degraded performance, download a newer compatible model or retrain your custom model with the current spaCy version. For more details and available updates, run: python -m spacy validate
  warnings.warn(warn_msg)


spacy.load will return a Language object containing all components and data needed to process text. This object is usually called nlp in the documentation and tutorials. Calling the nlp object on a string of text will return a processed Doc object with the text split into words and annotated.

Let’s get all identified entities and print their text, start position, end position, and type:

In [ ]:
doc = nlp(sample_transcription)
print("TEXT", "START", "END", "ENTITY TYPE")
for ent in doc.ents:
    print(ent.text, ent.start_char, ent.end_char, ent.label_)

TEXT START END ENTITY TYPE
iron-deficiency anemia 79 101 DISEASE
chronic blood loss 109 127 DISEASE
colitis 133 140 DISEASE
iron 203 207 CHEMICAL
colitis 286 293 DISEASE
hematoma 391 399 DISEASE
venous thrombosis 473 490 DISEASE
DVT 492 495 DISEASE
pelvic hematoma 742 757 DISEASE
vancomycin 781 791 CHEMICAL
infectious disease 873 891 DISEASE
improved.,PT 1348 1360 CHEMICAL
vitamin K 1503 1512 CHEMICAL
uric acid 1830 1839 CHEMICAL
bilirubin 1853 1862 CHEMICAL
Creatinine 1911 1921 CHEMICAL
creatinine 1951 1961 CHEMICAL
Folic acid 2079 2089 CHEMICAL
Iron 2103 2107 CHEMICAL
heparin 2322 2329 CHEMICAL
loperamide 2339 2349 CHEMICAL
niacin 2351 2357 CHEMICAL
pantoprazole 2359 2371 CHEMICAL
Diovan 2373 2379 CHEMICAL
Afrin 2381 2386 CHEMICAL
caspofungin 2400 2411 CHEMICAL
daptomycin 2413 2423 CHEMICAL
Ertapenem 2425 2434 CHEMICAL
fentanyl 2436 2444 CHEMICAL
morphine 2448 2456 CHEMICAL
pain 2464 2468 DISEASE
Compazine 2474 2483 CHEMICAL
Zofran 2487 2493 CHEMICAL
epistaxis 2629 2638 DISEASE
bleed

We can see the model correctly identified and label diseases such as iron-deficiency anemia, chronic blood loss, and many more. Lots of drugs were also identified, like vancomycin, Compazine, Zofran. The model can also identify common laboratory tested molecules such as creatinine, iron, bilirubin, uric acid.

Not everything is perfect though. See how some tokens are weirdly classified as chemicals, possibly due to punctuation marks:

improved.,PT 1348 1360 CHEMICAL
edema.,CARDIAC 3109 3123 CHEMICAL
Punctuation marks are usually removed in NLP preprocessing steps, but we can’t remove all of them here, otherwise, we may miss chemical names and would screw up quantities like drug dosage. However, we can solve this problem by removing the “.,” marks that appear to separate some sections of the transcription. It is important to know your data and your data’s domain to have a better comprehension of your results.

In [ ]:
import re

med_transcript_small['transcription'] = med_transcript_small['transcription'].apply(lambda x: re.sub('(\.,)', ". ", x))

In [ ]:
from spacy import displacy
displacy.render(doc[:100], style='ent', jupyter=True) # here I am printing just the first 100 tokens

Rule-based matching
Rule-based matching is similar to regular expressions, but spaCy’s rule-based matcher engines and components give you access to the tokens within the document and their relationships. We can combine this with the NER models to identify some pattern that includes our entities.

Let’s extract from the text the drug names and their reported dosages. This could be of real use to identify possible medication errors by checking if the dosages are in accordance with standards and guidelines.

In [ ]:
from spacy.matcher import Matcher

pattern = [{'ENT_TYPE':'CHEMICAL'}, {'LIKE_NUM': True}, {'IS_ASCII': True}]
matcher = Matcher(nlp.vocab)
matcher.add("DRUG_DOSE", [pattern])

The code above creates a pattern to identify a sequence of three tokens:

A token whose entity type is CHEMICAL (drug name)
A token that resembles a number (dosage)
A token that consists of ASCII characters (units, like mg or mL)
Then we initialize the Matcher with a vocabulary. The matcher must always share the same vocab with the documents it will operate on, so we use the nlp object vocab. We then add this pattern to the matcher and give it an ID.

Now we can loop through all transcriptions and extract the text matching this pattern:

In [ ]:
for transcription in med_transcript_small['transcription']:
    doc = nlp(transcription)
    matches = matcher(doc)
    for match_id, start, end in matches:
        string_id = nlp.vocab.strings[match_id]  # get string representation
        span = doc[start:end]  # the matched span
        print(string_id, start, end, span.text)

DRUG_DOSE 137 140 Xylocaine 20 mL
DRUG_DOSE 141 144 Marcaine 0.25%
DRUG_DOSE 208 211 Aspirin 81 mg
DRUG_DOSE 216 219 Spiriva 10 mcg
DRUG_DOSE 399 402 nifedipine 10 mg
DRUG_DOSE 226 229 aspirin one tablet
DRUG_DOSE 245 248 Warfarin 2.5 mg
DRUG_DOSE 67 70 Topamax 100 mg
DRUG_DOSE 73 76 Zoloft 100 mg
DRUG_DOSE 79 82 Abilify 5 mg
DRUG_DOSE 84 87 Motrin 800 mg
DRUG_DOSE 110 113 Xanax 1 mg
DRUG_DOSE 125 128 Colace 100 mg
DRUG_DOSE 168 171 Paxil 10 mg
DRUG_DOSE 175 178 Prednisone 20 mg
DRUG_DOSE 193 196 Metamucil one pack
DRUG_DOSE 207 210 Nexium 40 mg
DRUG_DOSE 1133 1136 Naprosyn one p.o
DRUG_DOSE 290 293 Lidocaine 1%
DRUG_DOSE 37 40 Altrua 60,
DRUG_DOSE 239 242 lidocaine 2%
DRUG_DOSE 97 100 Creatinine 1.3,
DRUG_DOSE 100 103 sodium 141,
DRUG_DOSE 104 107 potassium 4.0.
DRUG_DOSE 108 111 Calcium 8.6.
DRUG_DOSE 440 443 aspirin 81 mg
DRUG_DOSE 447 450 Lipitor 20 mg
DRUG_DOSE 454 457 Klonopin 0.5 mg
DRUG_DOSE 462 465 digoxin 0.125 mg
DRUG_DOSE 469 472 Lexapro 10 mg
DRUG_DOSE 484 487 Lasix 20 mg


We successfully extracted drugs and dosages, including different kinds of units like mg, mL, %, packs.

We used one NER model, but there lots of others and you should totally check them out. For instance, the en_ner_bionlp13cg_md model can identify anatomical parts, tissues, cell types, and more. Imagine what else you could do with that!